[Buffer] implement with customizable storages

twittemb opened the pull request:
Hi

This PR aims to explore an implementation without using actors as discussed here: https://forums.swift.org/t/swift-async-algorithms-buffer/61562

2 buffering algorithms are implemented:
- one that uses a queue to stack the upstream elements with unbounded/bufferingNewest/bufferingOldest policies
- one that uses a queue to stack the upstream elements until a limit is reached, suspending the upstream sequence in that case

The task that iterates over the upstream sequence is created on the first call to `next()`, unlike the original implementation where the task  is created when the iterator is instantiated.

The `BufferStorage` protocol is public and allows to implement a custom algorithm. The concrete implementations are not public (keeping a reference on their functions).

I have added a throughput measurement. I can see a slight improvement but not decisive.

I guess there is space for improvement ...

@phausler @FranzBusch

phausler commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,117 @@
+  public typealias AsyncIterator = Iterator
+
+  let base: Base
+  let send: @Sendable (Element) async -> Void
```
so using this route will end up preventing inline optimizations (which is perhaps one of the biggest performance boons swift has to offer). I'm sure that having it be part of the generic is likely much more performant

twittemb replied:
Hi,

I guess those 3 variants could benefit from inline optimizations right?

- using concrete types:
https://github.com/twittemb/swift-async-algorithms/blob/feature/buffer-concrete-types/Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

- using existential:
https://github.com/twittemb/swift-async-algorithms/blob/feature/buffer-existantial/Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

- using generics:
https://github.com/twittemb/swift-async-algorithms/blob/feature/buffer-generic-over-protocol/Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift

phausler replied:
I think either the generics or concrete types are the right way to go for perf. We could take the generics version and leave the construction with the storage off for now until we settle everything down. Adding it as a later addition.

twittemb replied:
Ok I’ll work on that tomorrow and publish the PR

phausler replied:
I was looking at that branch further - it seems to resolve some of the issues brought up. I am going to take a look at seeing if we can reduce the surface for the `BufferStorage` down such that it can be more of a simple interface.

twittemb replied:
Overall I'm happy with the result.
The perfs are good (+20% for the unbounded algorithm compared to the Actor version).
The AsyncStorage protocol requirements are simple.

What do you think ?

phausler commented on Sources/AsyncAlgorithms/Buffer/QueuedBufferStorage.swift:
```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+public enum QueuedBufferPolicy: Sendable {
+  case unbounded
+  case bufferingNewest(limit: UInt)
```
usually we use `Int` and fault on invalid values

twittemb replied:
modified.

phausler commented on Tests/AsyncAlgorithmsTests/TestBuffer.swift:
```diff
@@ -193,82 +106,196 @@ final class TestBuffer: XCTestCase {
+      $0.inputs[0].buffer(policy: .bufferingNewest(limit: 2))
       "X,,,,,,[45^]"
     }
   }
```
It would be good to adapt some cancellation tests

```suggestion
  }
  
  
  func test_buffer_limit_finishes_when_iteration_task_is_cancelled() async {
    let value: String? = "test"
    let source = Indefinite(value: value)
    let sequence = source.async.map {
      try? await Task.sleep(nanoseconds: 1_000_000)
      return $0
    }.buffer(limit: 3)
    let finished = expectation(description: "finished")
    let iterated = expectation(description: "iterated")
    let task = Task {
      var firstIteration = false
      for await _ in sequence {
        if !firstIteration {
          firstIteration = true
          iterated.fulfill()
        }
      }
      finished.fulfill()
    }
    // ensure the other task actually starts
    wait(for: [iterated], timeout: 1.0)
    // cancellation should ensure the loop finishes
    // without regards to the remaining underlying sequence
    task.cancel()
    wait(for: [finished], timeout: 1.0)
  }
  
  func test_buffer_policy_finishes_when_iteration_task_is_cancelled() async {
    let value: String? = "test"
    let source = Indefinite(value: value)
    let sequence = source.async.map {
      try? await Task.sleep(nanoseconds: 1_000_000)
      return $0
    }.buffer(policy: .unbounded)
    let finished = expectation(description: "finished")
    let iterated = expectation(description: "iterated")
    let task = Task {
      var firstIteration = false
      for await _ in sequence {
        if !firstIteration {
          firstIteration = true
          iterated.fulfill()
        }
      }
      finished.fulfill()
    }
    // ensure the other task actually starts
    wait(for: [iterated], timeout: 1.0)
    // cancellation should ensure the loop finishes
    // without regards to the remaining underlying sequence
    task.cancel()
    wait(for: [finished], timeout: 1.0)
  }
```

twittemb replied:
added.

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,117 @@
+
+public protocol BufferStorage<Element> {
+  associatedtype Element
+  @Sendable func send(element: Element) async
```
could you elaborate a bit please ?

phausler replied:
so in generic version you have:
```swift
@rethrows
public protocol BufferStorage<Element>: Sendable {
  associatedtype Element
  func send(element: Element) async
  func fail(error: Error)
  func finish()
  func next() async -> Result<Element?, Error>
}
```

Adding throws to the `send` method allows for sending to the buffer to fail; e.g. throwing when the buffer is full.

```swift
@rethrows
public protocol BufferStorage<Element>: Sendable {
  associatedtype Element
  func send(element: Element) async throws
  func fail(error: Error)
  func finish()
  func next() async -> Result<Element?, Error>
}
```

twittemb replied:
ok I see. I'll try to add that requirement. It should ha 0 impact on my code I think since the thrown exception will be catch within the `Task` and transmitted to the buffer as an error. It will make the `next` function throw when the consumer consumers that error.

As for this signature `func next() async throws -> Element?`, I'll give it a try.

I'm currently reworking a lit the code base. I have merged the `func finish()` and `func fail(error: Error)` functions into a single one `func finish(error: Error?)` and I've reworked the state machines to have less different states.

I guess I should have something by tomorrow now.

twittemb replied:
I've changed the function signatures to match your suggestion.

twittemb replied:
Overall I'm happy with the result.
The perfs are good (+20% for the unbounded algorithm compared to the Actor version).
The `AsyncStorage` protocol requirements are simple.

What do you think ?

twittemb commented on Sources/AsyncAlgorithms/Buffer/QueuedBufferStorage.swift:
```diff
@@ -0,0 +1,106 @@
+  func next() async -> Result<Element?, Error> {
+    guard !Task.isCancelled else {  return .success(nil) }
+
+    let isCancelled = ManagedCriticalState(false)
```
The cancellation is now handled at the state machine level.

phausler reviewed (approved):
The changes here look quite good to me; it definitely addresses the outstanding concerns about safety and also seems to resolve the long standing flakey test that occurred with buffering.

FranzBusch reviewed (changes requested):
Before diving into the actual implementation, I have a broader concern about the API we are going to expose here. With this new implementation we are exposing in my opinion some implementation details that are not necessary. More specifically we are exposing `BufferStorage`, `SuspendingBufferStorage` & `QueuedBufferStorage`.

Just from an API design perspective I think we should only expose this API

```swift
extension AsyncSequence where Self: Sendable {
    public func buffer(strategy: AsyncBufferSequence.Stragety ) -> AsyncBufferSequence<Self> {}
}

public struct AsyncBufferSequence<Base: AsyncSequence & Sendable>: AsyncSequence, Sendable {}

public struct AsyncBufferSequence {
    public struct Strategy {
        public static let unbounded: Strategy
        public static let limited: Strategy
        public static let limitedDroppingLatest: Strategy
        public static let limitedDroppingOldest: Strategy
    } 
}
```

The above API allows us to evolve the underlying type however we want without exposing any internals. Furthermore, it allows us to add whatever buffering strategy we come up with without breaking API.

twittemb commented:
> Before diving into the actual implementation, I have a broader concern about the API we are going to expose here. With this new implementation we are exposing in my opinion some implementation details that are not necessary. More specifically we are exposing `BufferStorage`, `SuspendingBufferStorage` & `QueuedBufferStorage`.
> 
> Just from an API design perspective I think we should only expose this API
> 
> ```swift
> extension AsyncSequence where Self: Sendable {
>     public func buffer(strategy: AsyncBufferSequence.Stragety ) -> AsyncBufferSequence<Self> {}
> }
> 
> public struct AsyncBufferSequence<Base: AsyncSequence & Sendable>: AsyncSequence, Sendable {}
> 
> public struct AsyncBufferSequence {
>     public struct Strategy {
>         public static let unbounded: Strategy
>         public static let limited: Strategy
>         public static let limitedDroppingLatest: Strategy
>         public static let limitedDroppingOldest: Strategy
>     } 
> }
> ```
> 
> The above API allows us to evolve the underlying type however we want without exposing any internals. Furthermore, it allows us to add whatever buffering strategy we come up with without breaking API.

I like the idea of a "unified" strategy. not sure of the naming though to make the distinction between the suspending vs non-suspending storages.

How would you unify the 2 algorithms under the "AsyncBufferSequence" banner ? 1 concrete type for each implementation (AsyncSuspendingBufferSequence and AsyncNonSuspendingBufferSequence) and then encapsulate them in a wrapper "AsyncBufferSequence" that just forwards the calls to the implementation?

FranzBusch commented:
> > The above API allows us to evolve the underlying type however we want without exposing any internals. Furthermore, it allows us to add whatever buffering strategy we come up with without breaking API.

I agree that naming can be bike shed. Though I don't think we have to call out in the name of the strategy that it suspends or not. We can do a lot with doc comments here and explain exactly how the various strategies pull from the upstream.


> How would you unify the 2 algorithms under the "AsyncBufferSequence" banner ? 1 concrete type per implementation (AsyncSuspendingBufferSequence and AsyncNonSuspendingBufferSequence) and then encapsulate them in a wrapper "AsyncBufferSequence" that just forwards the calls to the implementation?

I think we only need two concrete state machine & storage implementations and then have a private enum in the `AsyncBufferSequence` that looks like this:

```swift
enum Storage {
    case suspending
    case nonSuspending
}
``` 

However, I am questioning if we really need to implementation at all. IMO we should we be able to handle this all with a single state machine. We just need to have the same "latching" mechanism in the unstructured Task where the state machine dictates when to request the next value. In the case where we are non suspending we just always request the next.

> Also how would handle the different `Strategies` with the struct ? because that have different properties ...

I don't know if I understand you correctly but the spelled out version of the `Strategy` type would look like this:

public struct AsyncBufferSequence {
    public struct Strategy {
        internal enum _Strategy {
              case unbounded
              case limited
              ....
        }

        internal var _strategy: _Strategy

        public static let unbounded = Self(_strategy: .unbounded)
       ....
    } 
}

twittemb commented:
> However, I am questioning if we really need to implementation at all. IMO we should we be able to handle this all with a single state machine. We just need to have the same "latching" mechanism in the unstructured Task where the state machine dictates when to request the next value. In the case where we are non suspending we just always request the next.

I'm not even sure we need the "latching" mechanism. In fact we can see the non-suspending algorithm as a special case of the suspending one.

In the suspending version, the state machine checks if a suspension is needed before even suspending (like in the last implementation of the `AsyncChannel` in my PR). When the strategy is set to "non-suspending" then we can buffer the element is never suspend.

tcldr commented:
> In fact we can see the non-suspending algorithm as a special case of the suspending one.

I think that this will emerge as a recurring theme with SAA. 👍

twittemb commented:
> > > The above API allows us to evolve the underlying type however we want without exposing any internals. Furthermore, it allows us to add whatever buffering strategy we come up with without breaking API.
> 
> I agree that naming can be bike shed. Though I don't think we have to call out in the name of the strategy that it suspends or not. We can do a lot with doc comments here and explain exactly how the various strategies pull from the upstream.
> 
> > How would you unify the 2 algorithms under the "AsyncBufferSequence" banner ? 1 concrete type per implementation (AsyncSuspendingBufferSequence and AsyncNonSuspendingBufferSequence) and then encapsulate them in a wrapper "AsyncBufferSequence" that just forwards the calls to the implementation?
> 
> I think we only need two concrete state machine & storage implementations and then have a private enum in the `AsyncBufferSequence` that looks like this:
> 
> ```swift
> enum Storage {
>     case suspending
>     case nonSuspending
> }
> ```
> 
> However, I am questioning if we really need to implementation at all. IMO we should we be able to handle this all with a single state machine. We just need to have the same "latching" mechanism in the unstructured Task where the state machine dictates when to request the next value. In the case where we are non suspending we just always request the next.
> 
> > Also how would handle the different `Strategies` with the struct ? because that have different properties ...
> 
> I don't know if I understand you correctly but the spelled out version of the `Strategy` type would look like this:
> 
> public struct AsyncBufferSequence { public struct Strategy { internal enum _Strategy { case unbounded case limited .... }
> 
> ```
>     internal var _strategy: _Strategy
> 
>     public static let unbounded = Self(_strategy: .unbounded)
>    ....
> } 
> ```
> 
> }

Hi @FranzBusch I've pushed a unified version. I was not able to implement the `Strategy` with a struct as having `static` properties is not allowed in a generic type (assuming we keep this type nested in `AsyncBufferSequence`). Moreover could you tell me the benefit of having a struct + static properties when it is backed by an enum ?

FranzBusch commented:
> Hi @FranzBusch I've pushed a unified version. I was not able to implement the `Strategy` with a struct as having `static` properties is not allowed in a generic type (assuming we keep this type nested in `AsyncBufferSequence`). Moreover could you tell me the benefit of having a struct + static properties when it is backed by an enum ?

Ah right we are in a generic type here. Then we gotta move it to a separate top level type, e.g. `AsyncBufferSequenceStrategy`. The reason why we need to do this is that adding new enum cases is an API breaking change since enums can be exhaustively switched over. That means if you add a new enum case adopters will get compiler errors when they switch over it. The recommended pattern to avoid this right now is to use a struct + static lets.

twittemb commented:
> > Hi @FranzBusch I've pushed a unified version. I was not able to implement the `Strategy` with a struct as having `static` properties is not allowed in a generic type (assuming we keep this type nested in `AsyncBufferSequence`). Moreover could you tell me the benefit of having a struct + static properties when it is backed by an enum ?
> 
> Ah right we are in a generic type here. Then we gotta move it to a separate top level type, e.g. `AsyncBufferSequenceStrategy`. The reason why we need to do this is that adding new enum cases is an API breaking change since enums can be exhaustively switched over. That means if you add a new enum case adopters will get compiler errors when they switch over it. The recommended pattern to avoid this right now is to use a struct + static lets.

I've put the policy at the top level. I kept the `Policy` terminology because this is something that is already used for `AsyncStream` but I let you and @phausler give me the appropriate direction.

FranzBusch reviewed (commented):
This looks very good already. I think the `public` API is almost perfect.I left some more comments inline but I have one larger one. You merged the implementation of the bounded and unbounded buffer into a single state machine now. This results in a bit of friction since the unbounded cases can constantly request new elements from the upstream whereas the bounded case needs to suspend. With your current implementation the bounded case will request one too many element from the upstream since we suspend after calling `next()` on the upstream. We should correct that and only request as many elements as actually required.

I know we are going in circles a bit here but I really think we need the two separate `Storage` and `StateMachine`s that you had previously. I would call them `AsyncBoundedBufferStorage/StateMachine` and `AsyncUnboundedBufferStorage/StateMachine`. To still have this code performant I would recommend to change the iterator like this

```swift
  public struct Iterator: AsyncIteratorProtocol {
    private enum Backing {
      case bounded(AsyncBoundedBufferStorage)
      case unbounded(AsyncUnboundedBufferStorage)
    }
    
    private let backing: Backing

    init(_ policy: AsyncBufferSequencePolicy.Policy) {
      switch policy { ... }
    }
    
    public mutating func next() async re throws -> Element? {
        switch self.backing {
             case .bounded(let storage): 
               let result = await storage.next()
               return result.rethrow_get()

             case .unbounded(let storage): 
               let result = await storage.next()
               return result.rethrow_get()
        }
    }
```

Importantly here, the `Task` creation goes into the individual storage implementation since they need to do different things.

Let me know what you think! Happy to discuss this

twittemb commented:
> This looks very good already. I think the `public` API is almost perfect.I left some more comments inline but I have one larger one. You merged the implementation of the bounded and unbounded buffer into a single state machine now. This results in a bit of friction since the unbounded cases can constantly request new elements from the upstream whereas the bounded case needs to suspend. With your current implementation the bounded case will request one too many element from the upstream since we suspend after calling `next()` on the upstream. We should correct that and only request as many elements as actually required.
> 
> I know we are going in circles a bit here but I really think we need the two separate `Storage` and `StateMachine`s that you had previously. I would call them `AsyncBoundedBufferStorage/StateMachine` and `AsyncUnboundedBufferStorage/StateMachine`. To still have this code performant I would recommend to change the iterator like this
> 
> ```swift
>   public struct Iterator: AsyncIteratorProtocol {
>     private enum Backing {
>       case bounded(AsyncBoundedBufferStorage)
>       case unbounded(AsyncUnboundedBufferStorage)
>     }
>     
>     private let backing: Backing
> 
>     init(_ policy: AsyncBufferSequencePolicy.Policy) {
>       switch policy { ... }
>     }
>     
>     public mutating func next() async re throws -> Element? {
>         switch self.backing {
>              case .bounded(let storage): 
>                let result = await storage.next()
>                return result.rethrow_get()
> 
>              case .unbounded(let storage): 
>                let result = await storage.next()
>                return result.rethrow_get()
>         }
>     }
> ```
> 
> Importantly here, the `Task` creation goes into the individual storage implementation since they need to do different things.
> 
> Let me know what you think! Happy to discuss this

Thanks for the review. It means (for the `AsyncBoundedBufferStorage`) that we indeed need to "pre-suspend" before requesting the next element (latching mechanism that we already used) right ?

I guess we could still unify both implementations but it might affect the a little the perfs of the unbounded algorithm with the following sequence of events:

- do we need to suspend before next? (always false for `unbounded`, true or false for `bounded` depending on the buffer)
- eventually suspend until room in the buffer (never happens for `unbounded`)
- request next from upstream
- send the element to the state machine

I'm not trying to unify things at all cost :-)  just exploring if this is a satisfactory solution that could save us a bit of code. What do you think?

twittemb commented:
> > This looks very good already. I think the `public` API is almost perfect.I left some more comments inline but I have one larger one. You merged the implementation of the bounded and unbounded buffer into a single state machine now. This results in a bit of friction since the unbounded cases can constantly request new elements from the upstream whereas the bounded case needs to suspend. With your current implementation the bounded case will request one too many element from the upstream since we suspend after calling `next()` on the upstream. We should correct that and only request as many elements as actually required.
> > I know we are going in circles a bit here but I really think we need the two separate `Storage` and `StateMachine`s that you had previously. I would call them `AsyncBoundedBufferStorage/StateMachine` and `AsyncUnboundedBufferStorage/StateMachine`. To still have this code performant I would recommend to change the iterator like this
> > ```swift
> >   public struct Iterator: AsyncIteratorProtocol {
> >     private enum Backing {
> >       case bounded(AsyncBoundedBufferStorage)
> >       case unbounded(AsyncUnboundedBufferStorage)
> >     }
> >     
> >     private let backing: Backing
> > 
> >     init(_ policy: AsyncBufferSequencePolicy.Policy) {
> >       switch policy { ... }
> >     }
> >     
> >     public mutating func next() async re throws -> Element? {
> >         switch self.backing {
> >              case .bounded(let storage): 
> >                let result = await storage.next()
> >                return result.rethrow_get()
> > 
> >              case .unbounded(let storage): 
> >                let result = await storage.next()
> >                return result.rethrow_get()
> >         }
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Importantly here, the `Task` creation goes into the individual storage implementation since they need to do different things.
> > Let me know what you think! Happy to discuss this
> 
> Thanks for the review. It means (for the `AsyncBoundedBufferStorage`) that we indeed need to "pre-suspend" before requesting the next element (latching mechanism that we already used) right ?
> 
> I guess we could still unify both implementations but it might affect a bit the perfs of the unbounded algorithm with the following sequence of events:
> 
> * do we need to suspend before next? (always false for `unbounded`, true or false for `bounded` depending on the buffer)
> * eventually suspend until room in the buffer (never happens for `unbounded`)
> * request next from upstream
> * send the element to the state machine
> 
> I'm not trying to unify things at all cost :-) just exploring if this is a satisfactory solution that could save us a bit of code. What do you think?

After giving it a second though, it’s probably not worth trying to keep a single storage for both algorithms… it would make it more complex to understand because of the conditional behaviour whether it’s bounded or unbounded. I’ll get back to you soon with an implementation

twittemb commented:
> This looks very good already. I think the `public` API is almost perfect.I left some more comments inline but I have one larger one. You merged the implementation of the bounded and unbounded buffer into a single state machine now. This results in a bit of friction since the unbounded cases can constantly request new elements from the upstream whereas the bounded case needs to suspend. With your current implementation the bounded case will request one too many element from the upstream since we suspend after calling `next()` on the upstream. We should correct that and only request as many elements as actually required.
> 
> I know we are going in circles a bit here but I really think we need the two separate `Storage` and `StateMachine`s that you had previously. I would call them `AsyncBoundedBufferStorage/StateMachine` and `AsyncUnboundedBufferStorage/StateMachine`. To still have this code performant I would recommend to change the iterator like this
> 
> ```swift
>   public struct Iterator: AsyncIteratorProtocol {
>     private enum Backing {
>       case bounded(AsyncBoundedBufferStorage)
>       case unbounded(AsyncUnboundedBufferStorage)
>     }
>     
>     private let backing: Backing
> 
>     init(_ policy: AsyncBufferSequencePolicy.Policy) {
>       switch policy { ... }
>     }
>     
>     public mutating func next() async re throws -> Element? {
>         switch self.backing {
>              case .bounded(let storage): 
>                let result = await storage.next()
>                return result.rethrow_get()
> 
>              case .unbounded(let storage): 
>                let result = await storage.next()
>                return result.rethrow_get()
>         }
>     }
> ```
> 
> Importantly here, the `Task` creation goes into the individual storage implementation since they need to do different things.
> 
> Let me know what you think! Happy to discuss this

BTW @FranzBusch I'm still working on it :-) I've not abandoned ... just need to find some time.

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,114 @@
+  public static let unbounded = Self(policy: .unbounded)
+
+  /// A policy for buffering elements until the limit is reached. Then it will discard the oldest buffered elements.
+  public static func bufferingNewest(_ limit: Int) -> Self { Self(policy: .bufferingNewest(limit)) }
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,114 @@
+  public typealias AsyncIterator = Iterator
+
+  let base: Base
+  let storage: BufferStorage<Base>
```
done.

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,114 @@
+  /// A policy for buffering elements until the limit is reached. Then it will discard the oldest buffered elements.
+  public static func bufferingNewest(_ limit: Int) -> Self { Self(policy: .bufferingNewest(limit)) }
+
+  /// A policy for buffering elements until the limit is reached. Then it will discard the newest elements.
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,114 @@
+        self.task = Task { [base, storage] in
+          var iterator = base.makeAsyncIterator()
+          do {
+            while let element = try await iterator.next() {
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStateMachine.swift:
```diff
@@ -0,0 +1,276 @@
+  typealias SuspendedConsumer = UnsafeContinuation<Element?, Error>
+
+  private enum State {
+    case buffering(buffer: Deque<Result<Element?, Error>>, suspendedProducer: SuspendedProducer?, suspendedConsumer: SuspendedConsumer?)
```
not optional anymore.

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStateMachine.swift:
```diff
@@ -0,0 +1,276 @@
+  private enum State {
+    case buffering(buffer: Deque<Result<Element?, Error>>, suspendedProducer: SuspendedProducer?, suspendedConsumer: SuspendedConsumer?)
+    case modifying
+    case finished(buffer: Deque<Result<Element?, Error>>)
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStateMachine.swift:
```diff
@@ -0,0 +1,276 @@
+  private let policy: _AsyncBufferSequencePolicy
+
+  init(policy: _AsyncBufferSequencePolicy) {
+    assert(policy.isPositiveLimit, "The limit should be positive for the buffer operator to be efficient.")
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStateMachine.swift:
```diff
@@ -0,0 +1,276 @@
+
+  private enum State {
+    case buffering(buffer: Deque<Result<Element?, Error>>, suspendedProducer: SuspendedProducer?, suspendedConsumer: SuspendedConsumer?)
+    case modifying
```
on this one I'd like your input, I've noticed significant lower perfs without this state. what do you think ?

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,114 @@
+          var iterator = base.makeAsyncIterator()
+          do {
+            while let element = try await iterator.next() {
+              await storage.send(element: element)
```
I've completely refactored the iteration on the upstream.

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStorage.swift:
```diff
@@ -0,0 +1,136 @@
+//
+//===----------------------------------------------------------------------===//
+
+public struct BufferStorage<Base: AsyncSequence>: Sendable {
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStorage.swift:
```diff
@@ -0,0 +1,136 @@
+    self.stateMachine = ManagedCriticalState(BufferStateMachine<Base>(policy: policy))
+  }
+
+  public func send(element: Base.Element) async {
```
done

twittemb commented on Sources/AsyncAlgorithms/Buffer/BufferStateMachine.swift:
```diff
@@ -0,0 +1,276 @@
+    case resumeConsumerAndExit(continuation: SuspendedConsumer?, result: Result<Element?, Error>)
+  }
+
+  mutating func elementIsProduced(
```
done

twittemb commented:
> > This looks very good already. I think the `public` API is almost perfect.I left some more comments inline but I have one larger one. You merged the implementation of the bounded and unbounded buffer into a single state machine now. This results in a bit of friction since the unbounded cases can constantly request new elements from the upstream whereas the bounded case needs to suspend. With your current implementation the bounded case will request one too many element from the upstream since we suspend after calling `next()` on the upstream. We should correct that and only request as many elements as actually required.
> > I know we are going in circles a bit here but I really think we need the two separate `Storage` and `StateMachine`s that you had previously. I would call them `AsyncBoundedBufferStorage/StateMachine` and `AsyncUnboundedBufferStorage/StateMachine`. To still have this code performant I would recommend to change the iterator like this
> > ```swift
> >   public struct Iterator: AsyncIteratorProtocol {
> >     private enum Backing {
> >       case bounded(AsyncBoundedBufferStorage)
> >       case unbounded(AsyncUnboundedBufferStorage)
> >     }
> >     
> >     private let backing: Backing
> > 
> >     init(_ policy: AsyncBufferSequencePolicy.Policy) {
> >       switch policy { ... }
> >     }
> >     
> >     public mutating func next() async re throws -> Element? {
> >         switch self.backing {
> >              case .bounded(let storage): 
> >                let result = await storage.next()
> >                return result.rethrow_get()
> > 
> >              case .unbounded(let storage): 
> >                let result = await storage.next()
> >                return result.rethrow_get()
> >         }
> >     }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Importantly here, the `Task` creation goes into the individual storage implementation since they need to do different things.
> > Let me know what you think! Happy to discuss this
> 
> BTW @FranzBusch I'm still working on it :-) I've not abandoned ... just need to find some time.

@FranzBusch I've addressed all the comments so far, except the one on the `.modifying` state as I noticed an impact on the perfs.

That being said, I have an issue I can't resolve. From time to time the cancellation with the `.unbounded` policy is not applied.

You can test the throughput for this policy and from time to time it will fail because of an unwaited "finished" expectation. I've noticed the same behaviour with the test `test_given_a_buffered_with_unbounded_sequence_when_cancelling_consumer_then_the_iteration_finishes_and_the_base_is_cancelled` (it will happen if you run it repeatedly). It's like the system is so busy producing/consuming elements and copying the buffer that the cancellation is never catched. It does not happen with the other policies like bufferingLatest and bufferingOldest.

Honestly I can't figure out why ... I need your help on this.

Thanks.

twittemb commented:
> > > This looks very good already. I think the `public` API is almost perfect.I left some more comments inline but I have one larger one. You merged the implementation of the bounded and unbounded buffer into a single state machine now. This results in a bit of friction since the unbounded cases can constantly request new elements from the upstream whereas the bounded case needs to suspend. With your current implementation the bounded case will request one too many element from the upstream since we suspend after calling `next()` on the upstream. We should correct that and only request as many elements as actually required.
> > > I know we are going in circles a bit here but I really think we need the two separate `Storage` and `StateMachine`s that you had previously. I would call them `AsyncBoundedBufferStorage/StateMachine` and `AsyncUnboundedBufferStorage/StateMachine`. To still have this code performant I would recommend to change the iterator like this
> > > ```swift
> > >   public struct Iterator: AsyncIteratorProtocol {
> > >     private enum Backing {
> > >       case bounded(AsyncBoundedBufferStorage)
> > >       case unbounded(AsyncUnboundedBufferStorage)
> > >     }
> > >     
> > >     private let backing: Backing
> > > 
> > >     init(_ policy: AsyncBufferSequencePolicy.Policy) {
> > >       switch policy { ... }
> > >     }
> > >     
> > >     public mutating func next() async re throws -> Element? {
> > >         switch self.backing {
> > >              case .bounded(let storage): 
> > >                let result = await storage.next()
> > >                return result.rethrow_get()
> > > 
> > >              case .unbounded(let storage): 
> > >                let result = await storage.next()
> > >                return result.rethrow_get()
> > >         }
> > >     }
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Importantly here, the `Task` creation goes into the individual storage implementation since they need to do different things.
> > > Let me know what you think! Happy to discuss this
> > 
> > 
> > BTW @FranzBusch I'm still working on it :-) I've not abandoned ... just need to find some time.
> 
> @FranzBusch I've addressed all the comments so far, except the one on the `.modifying` state as I noticed an impact on the perfs.
> 
> That being said, I have an issue I can't resolve. From time to time the cancellation with the `.unbounded` policy is not applied.
> 
> You can test the throughput for this policy and from time to time it will fail because of an unwaited "finished" expectation. I've noticed the same behaviour with the test `test_given_a_buffered_with_unbounded_sequence_when_cancelling_consumer_then_the_iteration_finishes_and_the_base_is_cancelled` (it will happen if you run it repeatedly). It's like the system is so busy producing/consuming elements and copying the buffer that the cancellation is never catched. It does not happen with the other policies like bufferingLatest and bufferingOldest.
> 
> Honestly I can't figure out why ... I need your help on this.
> 
> Thanks.

Forget about it ... I found what was wrong. my cancellation management was not 100% good. it is fixed :-)

twittemb commented:
@phausler i guess you'd like to take a look at the new implementation as well since you approved a previous version.

FranzBusch commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,122 @@
+  /// - Parameter policy: A policy that drives the behaviour of the ``AsyncBufferSequence``
+  /// - Returns: An asynchronous sequence that buffers elements up to a given limit.
+  public func buffer(
+    policy: AsyncBufferSequence<Self>.Policy
```
I am wondering if this is the right way to do this. Didn't we last discuss to move the `Policy` into its standalone top level type called `AsyncBufferSequencePolicy`?. In the case where you just get a sequence and want to call `.buffer(policy: .unlimited)` this works great, but if you want to spell hold the policy somewhere. It is getting hard to spell it out. I am inclined to say we should love it to its own top level type.

phausler replied:
I wonder if the policy system is perhaps something we need for other things too. I agree that a buffering type being coupled with its policy seems appropriate (specifically the way it is currently written). But, if the policy system is general to any buffering mechanism then it should be at the top level.

FranzBusch commented on Sources/AsyncAlgorithms/Buffer/BoundedBufferStateMachine.swift:
```diff
@@ -0,0 +1,339 @@
+    self.limit = limit
+  }
+
+  var task: Task<Void, Never>? {
```
As far as I can see we are calling this from `deinit` only. In general, exposing state from the state machine through a computed prop should be considered an anti pattern. Since we are actually mutating state but the state machine isn't informed about it. Can we please create a method `storageDeinited` with an appropriate action. Same applied to the unbounded state machine.

twittemb replied:
done. (the behaviour is mutualised with next cancellation)

FranzBusch reviewed (commented):
Looks really great! Thanks for all the work again. I left two last comments and afterwards we should be good to merge this.

phausler commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,122 @@
+    /// A policy for buffering elements until the limit is reached.
+    /// Then consumption of the upstream `AsyncSequence` will be paused until elements are consumed from the buffer.
+    public static func bounded(_ limit: Int) -> Self {
+      precondition(limit > 0, "The limit should be positive for the buffer operator to be efficient.")
```
What happens if limit is 0? There are scenarios where that could be useful.

twittemb replied:
Interesting question. Before taking any actions I think we have to think about that.

The current behaviour will be:
- bounded policy: after the first call to next, the consumer will receive the first element and then the producer will suspend until the consumer request a next element. In the end it's like the buffer iterator has no effect here.

- bufferingNewest: after the first call to next, the consumer will receive the first element and then the producer will continue to emit elements. The buffer will stack only the latest emitted element (the previous one is discarded). When a new call to next is done then the penultimate element is consumed.

- bufferingOldest: after the first call to next, the consumer will receive the first element and then the producer will continue to emit elements. The buffer will never stack elements. We loose all the elements until the consumer request a next element.

I guess this is not ideal for the two last cases. Perhaps we should also suspend the producer in those cases?

What do you think @FranzBusch @phausler ?

twittemb replied:
@phausler @FranzBusch 

I've tried something. Considering that a limit <=0 should be equivalent to having no buffer, I've introduced a third `storageType` in the enum. This new case is `transparent` and just forwards the call to next to the upstream async sequence, as if there was no buffer at all. 

I've added some unit tests to make sure the upstream sequence timeline is not altered. I've also added comments in the doc and removed the preconditions.

Let me know what you think.

FranzBusch replied:
I think handling `0` with no buffer is fine; however, I still think we should `preconditionFailure` on `<0`.

twittemb replied:
done.

twittemb commented:
@FranzBusch @phausler thanks for the review, I’ll address the comments tomorrow.

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,122 @@
+  }
+}
+
+extension AsyncBufferSequence: Sendable where Base: Sendable { }
```
done.

twittemb commented on Sources/AsyncAlgorithms/Buffer/AsyncBufferSequence.swift:
```diff
@@ -0,0 +1,131 @@
+
+  /// A policy for buffering elements until the limit is reached.
+  /// After the limit is reached and a new element is produced by the upstream, the latest buffered element will be discarded.
+  /// If the limit is negative or zero then no buffering policy is applied.
```
oops ... I went too fast. It's fixed.
Thanks.

twittemb commented:
> Thanks for all the work. Looks great! One last nit but approved already

@phausler is it OK with you ?

phausler commented:
yea, this looks great! lets merge it

tcldr commented:
Nice one, @twittemb . Looks great.
